File: D2L1/RedAndGreen.hs
module RedAndGreen where 

best :: String -> Int -> Int
best [] _ = 0
best ('R':xs) reds = best xs (reds - 1)
best ('G':xs) reds = min (1 + (best xs reds)) reds

minPaints :: String -> Int
minPaints row = best row (length $ filter (== 'R') row)

File: D2L1/RedAndGreen.cpp
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

int minPaints(string const &row) {
    int n = row.size(), best = 0, reds = 0;
    for (int i = n - 1; i >= 0; --i) {
        switch (row[i]) {
        case 'R': 
            ++reds;
            break;
        case 'G':
            best = min(best + 1, reds);
        }
    }    
    return best;
}
